lib/xnvme_be_spdk_dev.c

```diff
@@ -689,6 +696,55 @@ xnvme_be_spdk_enumerate(const char *sys_uri, struct xnvme_opts *opts, xnvme_enum
 	return 0;
 }
 
+static int
+reconnect_ctrlr(struct spdk_nvme_ctrlr *ctrlr)
+{
+	int err;
+	XNVME_DEBUG("WARNING: attempting to reconnect controller");
+	err = spdk_nvme_ctrlr_reset(ctrlr);
+	if (err < 0) {
+		XNVME_DEBUG("FAILED: spdk_nvme_ctrlr_reset, err: %d", err);
+		return -EBUSY;
+	}
+	XNVME_DEBUG("INFO: controller reconnected");
+	return 0;
+}
+
+void
+_keep_alive_cb(void *ctx, const struct spdk_nvme_cpl *cpl)
+{
+	int *result = (int *)ctx;
+	*result = cpl->status.sct;
+	if (*result != 0) {
+		XNVME_DEBUG("WARNING: bad keep alive, result: %d", *result);
+	}
+}
+
+/**
+ * See if controller is able to handle a simple admin command if not then the ctrlr is in a bad
+ * state
+ */
+static int
+verify_ctrlr_ok(struct spdk_nvme_ctrlr *ctrlr)
+{
+	struct spdk_nvme_cmd cmd;
+	int cmd_result = -1;
+	int err;
+	cmd.opc = SPDK_NVME_OPC_KEEP_ALIVE;
+	spdk_nvme_ctrlr_cmd_admin_raw(ctrlr, &cmd, NULL, 0, _keep_alive_cb, &cmd_result);
+	while (cmd_result == -1) {
+		err = spdk_nvme_ctrlr_process_admin_completions(ctrlr);
+		if (err < 0) {
+			XNVME_DEBUG(
+				"WARNING: spdk_nvme_ctrlr_process_admin_completions failed, %d",
+				err);
+			cmd_result = err;
+			break;
+		}
+	}
+	return cmd_result;
```

Comment:
The `KEEP_ALIVE` command returns invalid opcode on my system.
I think it is better to use the `IDENTIFY` command instead.
With the code below:
```c
struct spdk_nvme_cmd cmd = {0};
int cmd_result = -1;
int err;

int buf_size = 4096;
void* buf;

buf = spdk_dma_malloc(buf_size, 0, NULL);
cmd.opc = SPDK_NVME_OPC_IDENTIFY;

spdk_nvme_ctrlr_cmd_admin_raw(ctrlr, &cmd, buf, buf_size, _keep_alive_cb, &cmd_result);
	while (cmd_result == -1) {
		err = spdk_nvme_ctrlr_process_admin_completions(ctrlr);
		if (err < 0) {
			XNVME_DEBUG(
				"WARNING: spdk_nvme_ctrlr_process_admin_completions failed, %d",
				err);
			cmd_result = err;
			break;
		}
	}

spdk_dma_free(buf);
return cmd_result;
```

Of course the `_keep_alive_cb` should be renamed to `_identify_cb`

Reply:
The following line is also needed
```c
cmd.cdw10_bits.identify.cns = SPDK_NVME_IDENTIFY_CTRLR;
```
I will run it over the weekend to see if the results are the same

Reply:
Yes, you're absolutely right. Nice catch!